### ISSUE TITLE
Support for dotnet?

### ISSUE THREAD

**Issue by Thaina** [lambci/docker-lambda#52]
I use C# for my lambda and want to test it in local. AWS point to here that we can test lambda locally by this docker. Would you pleased to add C# dotnet in this too?

**Comment by mhart** [comment]
Unfortunately this is a lot harder than the other languages, because replacing a single class in a .NET DLL is virtually impossible without having the original source.

I'm waiting on some support from AWS for this – I'll let you know when they get back to me

**Comment by joaoasrosa** [comment]
Hi @mhart,

Any updates on this?
Let me know if I can help.

Cheers

**Comment by mhart** [comment]
@joaoasrosa sure! How much do you know about disassembling and reassembling .NET DLLs? If you can replicate the same way I do it in Java, that'd be a huge help: https://github.com/lambci/docker-lambda/blob/master/java8/run/lambda-runtime-mock/build.sh#L7-L14

**Comment by jason-riddle** [comment]
@mhart Is there a `dotnet.tgz` file we can play around with in order to test dissembling and reassembling?

**Comment by jason-riddle** [comment]
Actually it looks like https://github.com/lambci/docker-lambda/pull/68 is attempting to address this issue.

**Comment by joaoasrosa** [comment]
@mhart and @jason-riddle if you do not mind I can take a look to #68 and try from there.

**Comment by romanlum** [comment]
The key diffrence between the .NET and the Java implementation ist that the .NET runtime is bootstraped from a NATIVE binary using native memory and native libraries...
That`s why i replaced the whole bootstrap in https://github.com/lambci/docker-lambda/pull/68

**Comment by mhart** [comment]
@romanlum it's actually still a compiled .NET binary though, and the libraries are .NET DLLs, except for the library that speaks to the AWS Lambda management layer, but that's the same with Java – that also uses a native library for that functionality, which is the only part I mock out.

Ideally the only part that is mocked out is just the part that communicates with the Lambda management layer – so that all of the rest of the code has the same behavior as it does on Lambda itself (eg, changing the way things are logged, or redirecting stdout or whatever).

**Comment by romanlum** [comment]
I don't think that the file /var/runtime/awslambda (which is the original entrypoint) is a compiled .net DLL.
It is a native application bootstraping the .net environment (calling /var/run/dotnet)

**Comment by romanlum** [comment]
Unfortunately we cannot use the original bootstrap.dll because there is a bug in the entrypoint which prevents us from replacing the lambda runtime as you did in java.

      if (!(str == "--lambda-runtime-override-assembly"))
              {
                if (!(str == "--lambda-runtime-override-type"))
                {
                  if (str == "--errorcode")
                  {
                    errorcode = s;
                  }
                  else
                  {
                    BootstrapEntrypoint.Usage(string.Format("Invalid argument '{0}'", (object) str));
                    return -1;
                  }
                }
                else
                  assemblyPath = s;
              }
              else
                assemblyPath = s;
            }
        
As you can see they are setting assemblyPath twice :(

**Comment by romanlum** [comment]
In my implementation I used the internal classes from the original bootstrap.dll so the loading and parsing of the data is the same as with the original runtime

**Comment by mhart** [comment]
@romanlum you are right, it does appear to be a binary compiled from C, not .NET – my apologies!

**Comment by mhart** [comment]
In dotnetcore1.0, the `/var/runtime/awslambda` binary usage is like this:
```
awslambda --netcore <coreclr/corefx> --runtime <bootstrap assemblies root> --task <user code root>
```
I haven't checked if it's the same for 2.0?

**Comment by romanlum** [comment]
its the same 
    bash-4.2# ./awslambda
    usage:
    awslambda
    --netcore <coreclr/corefx>
    --runtime <bootstrap assemblies root>
    --task <user code root>

**Comment by jason-riddle** [comment]
Is it possible to patch a dll file? Or would it be “easier” just to decompile the dll, fix the bug, and then recompile? Also, I’m guessing that the dll is signed by Amazon correct?

**Comment by romanlum** [comment]
The dll is not signed, but the native binary awslambda is calling the bootstrap.dll so you cannot add commandline arguments that easy .

If you look at my code in the pull request you can see that i am using the internal UserCodeLoader, so it is the same as if the bootstrap.dll would have been called.

**Comment by joaoasrosa** [comment]
@mhart there are further developments on this front? Anything where we can help?

**Comment by jason-riddle** [comment]
Update: https://github.com/lambci/docker-lambda/pull/68#issuecomment-367016977 has been merged which adds support for dotnet core 2.0.

**Comment by mhart** [comment]
Oh yeah, forgot we also had an issue open. .NET Core 2.0 is now supported!

I'm going to close this out. If there's really *a lot* of ppl asking for .NET Core 1.0 support, I could look at adding that, but I feel like it's pretty undesirable now that 2.0 is out.
